### PR TITLE
Separate `RemoteClusterActionType` from `ActionType`

### PR DESCRIPTION
**Pull request by DaveCTurner**
Remote-cluster actions are qualitatively different from local-cluster
actions: they translate into transport requests and must specify how to
deserialize the transport response, whereas local-cluster actions always
run against the local node. This commit introduces a new type,
`RemoteClusterActionType`, similar to `ActionType`, to reflect this
distinction.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/cluster/state/ClusterStateAction.java [hunk 1]
```diff
@@ -9,13 +9,19 @@
...
     public static final ClusterStateAction INSTANCE = new ClusterStateAction();
     public static final String NAME = "cluster:monitor/state";
+    public static final RemoteClusterActionType<ClusterStateResponse> REMOTE_TYPE = new RemoteClusterActionType<>(
+        NAME,
+        ClusterStateResponse::new
+    );
 
     private ClusterStateAction() {
-        super(NAME, ClusterStateResponse::new);
+        super(NAME, Writeable.Reader.localOnly());
```
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/state/ClusterStateAction.java:25]: Nit: I prefer the style in [PainlessExecuteAction.java](https://github.com/elastic/elasticsearch/pull/104611/files#diff-ff9ff526027147221fc0f045907c7df06ca43510d6f77ee52350417a0fae8982); I think it's a bit confusing to have ClusterStateAction be a "local" `ActionType` (constructed with `localOnly`) and have a `RemoteClusterActionType` field.

**Review by ldematte — APPROVED** [review]
Some minor (and completely optional) style comments, but LGTM!

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java [hunk 2]
```diff
@@ -115,7 +116,8 @@
 
 public class PainlessExecuteAction {
 
-    public static final ActionType<Response> INSTANCE = new ActionType<>("cluster:admin/scripts/painless/execute", Response::new);
+    public static final ActionType<Response> INSTANCE = ActionType.localOnly("cluster:admin/scripts/painless/execute");
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java:119]: Possibly, although then we would maybe want to rename `Client` into `LocalClusterClient` and so on. I'm ok with the implicit default being a local-cluster action, with remote-cluster actions being a fairly rare special case.

Also renaming things like that would be _very_ noisy, I wouldn't want to do it here.

File: server/src/main/java/org/elasticsearch/action/ActionType.java [hunk 3]
```diff
@@ -49,18 +49,16 @@ public static ActionType<ActionResponse.Empty> emptyResponse(String name) {
     }
 
     /**
-     * Construct an {@link ActionType} which callers can execute both on the local node (using {@link NodeClient}) and on a remote cluster
-     * (using a client obtained from {@link Client#getRemoteClusterClient}). If the action is only to be executed on the local cluster then
-     * declare it using {@link #localOnly} instead.
+     * Construct an {@link ActionType} and specify a method that can deserialize the response. This {@code responseReader} parameter is
+     * effectively unused. Use {@link #localOnly} instead.
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/ActionType.java:53]: No need to deprecate this, it's only used in this project and going away in a followup.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/state/ClusterStateAction.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/state/ClusterStateAction.java:25]: Yeah again there's lots of legacy in this area, I'd prefer not to subclass `ActionType` too (see #104139) but cleaning that up is a whole other marathon task.

Comment (ldematte) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/state/ClusterStateAction.java:25]: Yes, makes sense to do it separately (if ever)

**Pull request merged**